aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -860,6 +860,9 @@ Tensor zeros_like(
     const TensorOptions& options,
     c10::optional<c10::MemoryFormat> optional_memory_format) {
   if (options.layout() == kSparse && self.is_sparse()) {
```

Comment:
This is interesting as the check is that both the requested layout and the example tensor be sparse.

I tried creating a sparse tensor on Colab from a dense example:

```
t = torch.tensor((1, 2, 3))
torch.zeros_like(t, layout=torch.sparse_coo)
: Could not run 'aten::empty_strided' with arguments from the 'SparseCPU' backend. 'aten::empty_strided' is only available for these backends: [CPU, CUDA, BackendSelect, Autograd, Profiler, Tracer].
```

and a dense tensor from a sparse example:

```
t = torch.tensor((1, 2, 3)).to_sparse()
torch.zeros_like(t, layout=torch.strided, memory_format=torch.preserve_format)
: tensor([0, 0, 0])
```

and a sparse tensor from a sparse example:

```
t = torch.tensor((1, 2, 3)).to_sparse()
result = torch.zeros_like(t)
: tensor(indices=tensor([], size=(1, 0)),
       values=tensor([], size=(0,)),
       size=(3,), nnz=0, layout=torch.sparse_coo)
```

If the user doesn't specify a layout explicitly, what is layout set to? In particular, should this check `(has_layout() && layout == kSparse) || (!has_layout() && self.is_sparse())`?

Reply:
> If the user doesn't specify a layout explicitly, what is layout set to? 

I think, the `options` is derived from `self` if no layout is specified in python
```
# self: sparse 
# option layout: default
t = torch.tensor((1, 2, 3)).to_sparse()
torch.zeros_like(t)
> tensor(indices=tensor([], size=(1, 0)),
       values=tensor([], size=(0,)),
       size=(3,), nnz=0, layout=torch.sparse_coo)

# self: dense 
# option layout: default
t = torch.tensor((1, 2, 3))
torch.zeros_like(t)
> tensor([0, 0, 0])
```
this is due to how the code is generated : 
```
# torch/csrc/autograd/generated/python_torch_functions.cpp
#    │21490     // aten::zeros_like(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
#    │21491     auto self = _r.tensor(0);
#    │21492     const auto options = TensorOptions()
#    │21493         .dtype(_r.scalartypeWithDefault(2, self.scalar_type()))                                                                                                                #    │21494         .device(_r.deviceWithDefault(4, self.device()))                                                                                                                        #    │21495         .layout(_r.layoutWithDefault(3, layout_from_backend(self.options().backend())))                                                         #    │21496         .requires_grad(_r.toBool(6))                                                                                                                                           #    │21497         .pinned_memory(_r.toBool(5));                                                                                                                                          #    │21498     torch::utils::maybe_initialize_cuda(options);                                                                                                                              #    │21499     auto dispatch_zeros_like = [](const Tensor & self, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) -> Tensor {
#    │21500       pybind11::gil_scoped_release no_gil;                                                                                                                                     #    │21501       return torch::zeros_like(self, options, memory_format);                                                                                                                  #    │21502     };                                                                                                                                                                         #   >│21503     return wrap(dispatch_zeros_like(self, options, _r.memoryformatOptional(1)));                                                                                               #    │21504     Py_RETURN_NONE;                                                                                                                                                            #    │21505     END_HANDLE_TH_ERRORS                                                                                                                                                       #    │21506   }
```
So, maybe we can keep the current check.